### PR TITLE
App migration steps need to push feedback to the user ....

### PR DESCRIPTION
**Pull request by DeepDiver1975**
Pre and post migration repair steps can be defined in the info.xml of an app

bildschirmfoto von 2016-04-19 15-34-04
# ToDos:
- [ ] Add a background repair step which can be executed after upgrade during live operations
- [ ] Add a progress based feedback for longer running repair steps
- [ ] cleanup emitter vs. EventDispatcher (tech dept)

### CODE REVIEW

**Comment by nickvergessen** [comment]
PHP Fatal error:  Call to undefined method OC\HTTPHelper::linkToDocs() in /ssd/jenkins/workspace/ocs-api-integration-tests-ci/lib/private/App/InfoParser.php on line 80

**Comment by nickvergessen** [comment]
`OC_App::getAppInfo()` needs adjustment

**Comment by DeepDiver1975** [comment]
@PVince81 @schiesbn @nickvergessen @icewind1991 @blizzz @rullzer @MorrisJobke @karlitschek @LukasReschke 

Please review and leave you comments :wink: THX

**Comment by karlitschek** [comment]
nice approach!! 👍

File: lib/private/repair.php [hunk 1]
```diff
@@ -91,9 +94,21 @@ public function run() {
...
 	 *
-	 * @param RepairStep $repairStep repair step
+	 * @param RepairStep|string $repairStep repair step
 	 */
 	public function addStep($repairStep) {
+		if (is_string($repairStep)) {
+			if (class_exists($repairStep)) {
+				$s = new $repairStep();
+				if ($s instanceof \OC\RepairStep) {
+					$this->addStep($s);
```
Comment (icewind1991) [inline lib/private/repair.php]: This should either return, `$this->repairSteps[] = $repairStep;` should be moved to an `else` block or you can change this line with `$repairStep = $s`.

As it currently is it adds both the class name as well as an instance to `$this->repairSteps` from what I can tell

Comment (DeepDiver1975) [inline lib/private/repair.php]: This will add the created instance to the array

Comment (icewind1991) [inline lib/private/repair.php]: The recursive call will add the instance, but the current call to the function (with `$repairStep` a string) will continue and add the string to the array

Comment (DeepDiver1975) [inline lib/private/repair.php]: Oh - right! THX

File: lib/private/App/InfoParser.php [hunk 2]
```diff
@@ -79,12 +71,21 @@ public function parse($file) {
 		if (!array_key_exists('types', $array)) {
 			$array['types'] = array();
 		}
+		if (!array_key_exists('repair-steps', $array)) {
+			$array['repair-steps'] = array();
```
Comment (MorrisJobke) [inline lib/private/App/InfoParser.php]: Old array syntax 🙈

**Comment by DeepDiver1975** [comment]
@icewind1991 @MorrisJobke comments addressed - THX

**Comment by DeepDiver1975** [comment]
Should we deprecate update.php in favor of this new approach?
Like add a check on the occ app checker command?

@karlitschek @nickvergessen @MorrisJobke @PVince81

**Comment by karlitschek** [comment]
I think it is ok to have both in parallel for a release or two. Then we see how the new system works and then we can deprecate the old one.

**Comment by DeepDiver1975** [comment]
> I think it is ok to have both in parallel for a release or two.

I don't say kill it right away - but because we want to have proper migration feedback we need to enforce the new way. Deprecation will allow us to kill it in 2019 - just saying :wink:

File: lib/private/app.php [hunk 3]
```diff
@@ -1174,6 +1175,34 @@ public static function updateApp($appId) {
 
 	/**
 	 * @param string $appId
+	 * @param array | string[] $steps
```
Comment (icewind1991) [inline lib/private/app.php]: `RepairStep[]|string[]` ?

File: lib/private/updater.php [hunk 4]
```diff
@@ -473,5 +475,31 @@ private function upgradeAppStoreApps(array $disabledApps) {
...
+
+	/**
+	 * Forward messages emitted by the repair routine
+	 *
+	 * @param Repair $repair repair routine
+	 */
+	private function emitRepairEvents() {
+		$dispatcher = \OC::$server->getEventDispatcher();
+		$dispatcher->addListener('\OC\Repair::warning', function ($event) {
+			/** @var GenericEvent $event*/
```
Comment (icewind1991) [inline lib/private/updater.php]: Type hint the function instead?

**Comment by icewind1991** [comment]
Code looks good otherwise

:+1: for deprecating update.php

File: lib/private/app.php (hunk 3 above)
Comment (DeepDiver1975) [inline lib/private/app.php]: should be only string[]

File: lib/private/repair.php [hunk 5]
```diff
@@ -46,20 +46,24 @@
...
+	private $dispatcher;
 
 	/**
 	 * Creates a new repair step runner
 	 *
-	 * @param array $repairSteps array of RepairStep instances
+	 * @param RepairStep[] $repairSteps array of RepairStep instances
 	 */
-	public function __construct($repairSteps = array()) {
+	public function __construct($repairSteps = [], EventDispatcher $dispatcher = null) {
```
Comment (LukasReschke) [inline lib/private/repair.php]: PHPDoc lies now? 😉

**Comment by LukasReschke** [comment]
I like the approach. 👍  besides some small nitpicks in the code.

**Comment by DeepDiver1975** [comment]
Open ToDos will be addressed in follow up prs

**Pull request merged**